File: src/main/scala/com/usercase/resume/input/RawFileRecordReader.scala
package com.usercase.resume.input

import java.io.ByteArrayOutputStream
import java.nio.charset.Charset
import java.util.zip.{ZipEntry, ZipInputStream}

import org.apache.hadoop.fs.FSDataInputStream
import org.apache.hadoop.io.{BytesWritable, Text}
import org.apache.hadoop.mapreduce.lib.input.FileSplit
import org.apache.hadoop.mapreduce.{InputSplit, RecordReader, TaskAttemptContext}

import scala.util.control.Breaks

/**
  * Created by devops on 2017/3/28.
  */
class RawFileRecordReader extends RecordReader[Text, BytesWritable] {

  private var fsin: FSDataInputStream = null

  private var zip : ZipInputStream = null

  private var currentKey: Text = null

  private var currentValue: BytesWritable = null

  private var isFinished = false


  override def getCurrentKey: Text = {
    currentKey
  }

  override def getProgress: Float = {

    return if(isFinished) 1f else 0f
  }

  override def nextKeyValue(): Boolean = {

    var bos: ByteArrayOutputStream  = null

    if(isFinished) {

      close()

      return false

    } else if(zip == null) { // doc,docx,txt,pdf

      // println("read other type file")

      bos = new ByteArrayOutputStream()

      val tmp = new Array[Byte](8192)

      val break = new Breaks

      break.breakable {

        while(true) {

          var bytesRead = 0

          try {
            bytesRead = fsin.read(tmp, 0, 8192)
          } catch {
            case e: Exception =>
              return false
          }

          if(bytesRead > 0) {
            bos.write(tmp,0,bytesRead)
          } else  break.break()
        }
      }

      isFinished = true

    } else { // ZIP

      // println("read zip file")

      bos = new ByteArrayOutputStream()

      var entry: ZipEntry = null

       try {
         entry = zip.getNextEntry
       } catch {
         case e: Exception =>
           return false
       }

      currentKey = new Text(entry.getName)

      val tmp = new Array[Byte](8192)

      val break = new Breaks

      break.breakable {

        while(true) {

          var bytesRead = 0

          try {
            bytesRead = zip.read(tmp, 0, 8192)
          } catch {
            case e: Exception =>
              return false
          }

          if(bytesRead > 0) {
            bos.write(tmp,0,bytesRead)
          } else  break.break()
        }
      }

      zip.closeEntry()

    }

    bos.flush()

    currentValue = new BytesWritable(bos.toByteArray())

    isFinished = true

    // println("currentvalue:" + currentValue)

    return true

  }

  override def getCurrentValue: BytesWritable = {
    currentValue
  }

  override def initialize(inputSplit: InputSplit, taskAttemptContext: TaskAttemptContext): Unit = {

    if (fsin == null) {

      val fileSplit = inputSplit.asInstanceOf[FileSplit]
      val conf = taskAttemptContext.getConfiguration
      val path = fileSplit.getPath
      val fs = path.getFileSystem(conf)

      fsin = fs.open(path)

      if (path.getName().endsWith(".zip")) {

        zip = new ZipInputStream(fsin, Charset.forName("GBK"))

        println("zip file")

      } else {
        currentKey = new Text(path.getName())

        // println("other type file")

      }

    }

  }

  override def close(): Unit = {

    try {
      zip.close()
    } catch  {
      case  e: Exception =>

    }

    try {
      fsin.close()
    } catch  {
      case  e: Exception =>

    }

  }
}

File: src/main/java-code/com/jbgfurfeature/hadoop/RawFileRecordReader.java
package com.jbgfurfeature.hadoop;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FSDataInputStream;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.BytesWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.InputSplit;
import org.apache.hadoop.mapreduce.RecordReader;
import org.apache.hadoop.mapreduce.TaskAttemptContext;
import org.apache.hadoop.mapreduce.lib.input.CombineFileSplit;
import org.apache.hadoop.mapreduce.lib.input.FileSplit;

import java.io.ByteArrayOutputStream;
import java.io.EOFException;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.zip.ZipEntry;
import java.util.zip.ZipException;
import java.util.zip.ZipInputStream;

/**
 * This RecordReader implementation extracts individual files from a ZIP
 * file and hands them over to the Mapper. The "key" is the decompressed
 * file name, the "value" is the file contents.
 */
public class RawFileRecordReader
    extends RecordReader<Text, BytesWritable> {
  /**
   * InputStream used to read the ZIP file from the FileSystem
   */
  private FSDataInputStream fsin;

  /**
   * ZIP file parser/decompresser
   */
  private ZipInputStream zip;

  /**
   * Uncompressed file name
   */
  private Text currentKey;

  /**
   * Uncompressed file contents
   */
  private BytesWritable currentValue;

  /**
   * Used to indicate progress
   */
  private boolean isFinished = false;


  public RawFileRecordReader() {

  }

  public RawFileRecordReader(CombineFileSplit split, TaskAttemptContext context, Integer index)
      throws IOException {
    Path path = split.getPath(index);
    FileSystem fs = path.getFileSystem(context.getConfiguration());
    // Open the stream
    fsin = fs.open(path);
    if (path.getName().endsWith(".zip")) {
//            zip = new ZipInputStream(fsin);
      zip = new ZipInputStream(fsin, Charset.forName("GBK"));
    } else {
      currentKey = new Text(path.getName());
    }

  }

  /**
   * Initialise and open the ZIP file from the FileSystem
   */
  @Override
  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext)
      throws IOException, InterruptedException {
    if (fsin == null) {
      FileSplit split = (FileSplit) inputSplit;
      Configuration conf = taskAttemptContext.getConfiguration();
      Path path = split.getPath();
      FileSystem fs = path.getFileSystem(conf);

      // Open the stream
      fsin = fs.open(path);
      if (path.getName().endsWith(".zip")) {
//                zip = new ZipInputStream(fsin);
        zip = new ZipInputStream(fsin, Charset.forName("GBK"));
      } else {
        currentKey = new Text(path.getName());
      }
    }
  }

  /**
   * This is where the magic happens, each ZipEntry is decompressed and
   * readied for the Mapper. The contents of each file is held *in memory*
   * in a BytesWritable object.
   *
   * If the ZipFileInputFormat has been set to Lenient (not the default),
   * certain exceptions will be gracefully ignored to prevent a larger job
   * from failing.
   */
  @Override
  @SuppressWarnings("IllegalCatch")
  public boolean nextKeyValue()
      throws IOException, InterruptedException {
    ByteArrayOutputStream bos = null;
    if (isFinished) {
      try {
        zip.close();
      } catch (Exception ignore) {
      }
      try {
        fsin.close();
      } catch (Exception ignore) {
      }
      return false;
    } else if (zip == null) {
      bos = new ByteArrayOutputStream();
      // currentKey

      // Read the file contents
      byte[] temp = new byte[8192];
      while (true) {
        int bytesRead = 0;
        try {
          bytesRead = fsin.read(temp, 0, 8192);
        } catch (EOFException e) {
          if (!RawFileInputFormat.getLenient()) {
            throw e;
          }
          e.printStackTrace();
          return false;
        }
        if (bytesRead > 0) {
          bos.write(temp, 0, bytesRead);
        } else {
          break;
        }
      }
      isFinished = true;
    } else {
      bos = new ByteArrayOutputStream();
      ZipEntry entry = null;
      try {
        entry = zip.getNextEntry();
      } catch (ZipException e) {
        if (!RawFileInputFormat.getLenient()) {
          throw e;
        }
      } catch (IllegalArgumentException iae) {
        isFinished = true;
        return false;
      }

      // Sanity check
      if (entry == null) {
        isFinished = true;
        return false;
      }

      // Filename
      currentKey = new Text(entry.getName());

      // Read the file contents
      byte[] temp = new byte[8192];
      while (true) {
        int bytesRead = 0;
        try {
          bytesRead = zip.read(temp, 0, 8192);
        } catch (EOFException e) {
          if (!RawFileInputFormat.getLenient()) {
            throw e;
          }
          return false;
        }
        if (bytesRead > 0) {
          bos.write(temp, 0, bytesRead);
        } else {
          break;
        }
      }
      zip.closeEntry();
    }

    // Uncompressed contents
    currentValue = new BytesWritable(bos.toByteArray());
    return true;

  }

  /**
   * Rather than calculating progress, we just keep it simple
   */
  @Override
  public float getProgress()
      throws IOException, InterruptedException {
    return isFinished ? 1 : 0;
  }

  /**
   * Returns the current key (name of the zipped file)
   */
  @Override
  public Text getCurrentKey()
      throws IOException, InterruptedException {
    return currentKey;
  }

  /**
   * Returns the current value (contents of the zipped file)
   */
  @Override
  public BytesWritable getCurrentValue()
      throws IOException, InterruptedException {
    return currentValue;
  }

  /**
   * Close quietly, ignoring any exceptions
   */
  @Override
  @SuppressWarnings("IllegalCatch")
  public void close()
      throws IOException {
    try {
      zip.close();
    } catch (Exception ignore) {
    }
    try {
      fsin.close();
    } catch (Exception ignore) {
    }
  }
}
